### PR TITLE
feat: parallelize the chain validator

### LINKED ISSUE #2905: Optimize and parallelize the chain validator
**Issue summary**

Computing block state trees from scratch is inherently a sequential operation and cannot be parallelized. However, validating state trees that have already been computed can be done in parallel. This kind of validation is precisely what the `validate_chain` function does. As of writing this issue, the `validate_chain` function does not validate in parallel and has several other performance problems. In the past, the performance of the validator has not been a priority, but in the future, we would like to re-validate the entirety of the calibnet chain and large parts of the mainnet chain.

- [x] Modify `validate_chain` to use `N` threads for validation.
- [ ] Modify `validate_chain` to avoid all writes to the database. Validation only requires read access to the database.
- [ ] Make sure that no bottlenecks prevent using all available cores.
- [ ] Investigate whether `compute_tipset_state` or `apply_blocks` can be modified for better performance.

Note: Use the full snapshot for calibnet when testing. It contains all the state trees for the first 450000 epochs.

**Other information and links**

A full-snapshot for calibnet can be downloaded here:
https://forest-snapshots.fra1.cdn.digitaloceanspaces.com/debug/filecoin_full_calibnet_2023-04-07_450000.car

https://github.com/ChainSafe/forest/blob/9e6450c4496e0fdc0144a8ff7884d8f4b68af5b0/blockchain/state_manager/src/lib.rs#L1206

https://github.com/ChainSafe/forest/blob/9e6450c4496e0fdc0144a8ff7884d8f4b68af5b0/blockchain/state_manager/src/lib.rs#L736-L739

https://github.com/ChainSafe/forest/blob/9e6450c4496e0fdc0144a8ff7884d8f4b68af5b0/blockchain/state_manager/src/lib.rs#L360-L364

### PR DESCRIPTION
**Pull request by aatifsyed**
## Summary of changes
- added some documentation
- refactor chain validation so it's actually understandable and lazy
- make it parallel
- add some better tracing infra (I'll add more)

I've still been seeing cross-thread waits, see documentation.

image

Closes
#2905 

## Change checklist

- [ ] I have performed a self-review of my own code,
- [ ] I have made corresponding changes to the documentation,
- [ ] I have added tests that prove my fix is effective or that my feature works (if possible),
- [ ] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/state_manager/mod.rs [hunk 1]
```diff
@@ -1230,53 +1250,81 @@ where
...
+            .par_bridge()
+            .try_for_each(|(child, parent)| {
+                info!(height = parent.epoch(), "compute parent state");
+                let (actual_state, actual_receipt) = self
+                    .compute_tipset_state_blocking(parent, NO_CALLBACK)
+                    .context("couldn't compute tipset state")?;
+                let expected_receipt = child
+                    .blocks()
+                    .first()
+                    .context("child has no receipts")?
```
Comment (lemmih) [inline src/state_manager/mod.rs:1310]:
```suggestion
                    .min_ticket_block()
```

File: src/state_manager/mod.rs [hunk 2]
```diff
@@ -1230,53 +1250,81 @@ where
...
-        Ok(())
+    /// Validates all tipsets at epoch `start..=end` behind the heaviest tipset.
+    ///
+    /// This spawns [`rayon::current_num_threads`] threads to do the compute-heavy work
+    /// of tipset validation.
+    ///
+    /// # What is validation?
+    /// Every state transition returns a new _state root_, which is typically retained in e.g snapshots.
+    /// For "full" snapshots, all state roots are retained.
+    /// For standard snapshots, the last 1800 or so state roots are retained.
```
Comment (lemmih) [inline src/state_manager/mod.rs:1261]:
```suggestion
    /// For standard snapshots, the last 2000 state roots are retained.
```

File: src/state_manager/mod.rs [hunk 3]
```diff
@@ -752,8 +753,27 @@ where
...
-    /// Performs a state transition, and returns the state and receipt root of
-    /// the transition.
+    /// Conceptually, a [`Tipset`] consists of _blocks_ which share an _epoch_.
+    /// Each _block_ contains _messages_, which are executed by the _Filecoin Virtual Machine_
+    /// in a _transaction_.
+    ///
+    /// VM transaction execution essentially looks like this:
+    /// ```text
+    /// previous_state * transaction = new_state
+    /// ```
```
Comment (lemmih) [inline src/state_manager/mod.rs:763]: This picture might be too simplistic. You cannot get the next state if you only have the previous state and a transaction. Transactions may run query states as far back as 1800 epochs. So we need 1800 states and a transaction to compute the next state.

**Review by lemmih — APPROVED** [review]
Great code and great documentation.

File: src/daemon/mod.rs [hunk 4]
```diff
@@ -438,10 +437,15 @@ pub(super) async fn start(
...
-    if config.client.snapshot {
-        if let Some(validate_height) = config.client.snapshot_height {
-            ensure_params_downloaded().await?;
-            validate_chain(&state_manager, validate_height).await?;
+    if let (true, Some(validate_from)) = (config.client.snapshot, config.client.snapshot_height) {
+        /// We've been provided a snapshot and asked to validate it
+        ensure_params_downloaded().await?;
+        let current_height = state_manager.chain_store().heaviest_tipset().epoch();
+        assert!(current_height.is_positive());
+        match validate_from.is_negative() {
```
Comment (LesnyRumcajs) [inline src/daemon/mod.rs:445]: What is wrong with a simple `if-else`?

File: src/state_manager/mod.rs [hunk 5]
```diff
@@ -1230,53 +1252,77 @@ where
...
-            last_receipt = msg_root;
-        }
-        Ok(())
+    /// Validates all tipsets at epoch `start..=end` behind the heaviest tipset.
+    ///
+    /// This spawns [`rayon::current_num_threads`] threads to do the compute-heavy work
+    /// of tipset validation.
+    ///
+    /// # What is validation?
+    /// Every state transition returns a new _state root_, which is typically retained in e.g snapshots.
```
Comment (LesnyRumcajs) [inline src/state_manager/mod.rs:1261]:
```suggestion
    /// Every state transition returns a new _state root_, which is typically retained in, e.g., snapshots.
```

File: src/state_manager/mod.rs [hunk 6]
```diff
@@ -1230,53 +1252,77 @@ where
...
+    /// This function takes advantage of that fact to validate tipsets:
+    /// - `tipset[N]` claims that `receipt_root[N-1]` should be `0xDEADBEEF`
+    /// - find `tipset[N-1]`, and perform its state transition to get the actual `receipt_root`
+    /// - assert that they match
+    ///
+    /// See [`Self::compute_tipset_state_blocking`] for an explanation of state transitions.
+    ///
+    /// # Known issues
+    /// This function is blocking, but we do observe threads waiting and synchronizing.
+    /// This is suspected to be due to the VM's use of the chain store, or maybe JIT-ing.
```
Comment (LesnyRumcajs) [inline src/state_manager/mod.rs:1277]: What kind of JIT?

File: src/state_manager/mod.rs [hunk 7]
```diff
@@ -1230,53 +1252,77 @@ where
...
+            .tuple_windows()
+            .par_bridge()
+            .try_for_each(|(child, parent)| {
+                info!(height = parent.epoch(), "compute parent state");
+                let (actual_state, actual_receipt) = self
+                    .compute_tipset_state_blocking(parent, NO_CALLBACK)
+                    .context("couldn't compute tipset state")?;
+                let expected_receipt = child.min_ticket_block().message_receipts();
+                let expected_state = child.parent_state();
+                match (expected_state, expected_receipt) == (&actual_state, &actual_receipt) {
```
Comment (LesnyRumcajs) [inline src/state_manager/mod.rs:1311]: Once again, why not a simple `if-else`? What is the advantage of this verbosity and an increased indentation level?

**Review by LesnyRumcajs — APPROVED** [review]

File: src/cli_shared/logger/mod.rs [hunk 8]
```diff
@@ -57,17 +58,29 @@ pub fn setup_logger(
         None
     };
 
+    // Go to <https://ui.perfetto.dev> to browse trace files.
```
[the remark this answers is not in the document]
Comment (aatifsyed) [reply src/cli_shared/logger/mod.rs:61]: https://github.com/ChainSafe/forest/issues/3070

File: src/daemon/mod.rs (hunk 4 above)
Comment (aatifsyed) [reply src/daemon/mod.rs:445]: Same as here:
https://github.com/ChainSafe/forest/pull/2893#discussion_r1220358284

File: src/state_manager/mod.rs (hunk 7 above)
Comment (aatifsyed) [reply src/state_manager/mod.rs:1311]: Same as here:

https://github.com/ChainSafe/forest/pull/2893#discussion_r1220358284

File: src/state_manager/mod.rs (hunk 6 above)
Comment (aatifsyed) [reply src/state_manager/mod.rs:1277]: Not sure, taken from this comment. I'll reword
https://chainsafeio.slack.com/archives/C0144LHN732/p1687512425630809?thread_ts=1687471388.702309&cid=C0144LHN732

**Pull request merged**